_posts/2025-05-05-Exposing-WASM-binaries-as-MCP-tools.adoc

```diff
@@ -0,0 +1,122 @@
+    <wasm-tool name="my-wasm-tool" path="path/relative/to/module/root/binary.wasm"/>
+    <!-- Other binary definitions -->
+</subsystem>
+----
+
+== Next Step: CDI Integration
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI.
+The first way you can use them is to get them by name as an `org.wildfly.wasm.api.WasmInvoker` using the `@WasmTool` annotation.
+This very basic interface wraps the call to the binary, passing the parameters as a byte array and getting the result as a byte array.
+
+[source,java]
+----
+
+    @Inject
+    @WasmTool(value = "my-wasm-tool")
+    org.wildfly.wasm.api.WasmInvoker greet;
+
+    public String greet(String person) {
+        // "greet" is the name of the exported method of the WASM module
+        byte[] output = greet.call("greet", person.getBytes(StandardCharsets.UTF_8));
+        return new String(output, StandardCharsets.UTF_8);
+    }
+
+----
+
+A second way is to use the `@WasmToolService` over an interface to create a dynamic proxy wrapping the call to the WASM module and providing a way to serialize and deserialize your arguments to the byte array required by the invoker.
+
+[source,java]
+----
+@WasmToolService(wasmToolConfigurationName = "pizza", wasmMethodName = "retrievePizzeriaAddresses", argumentSerializer = CityJsonSerializer.class)
+public interface Pizza {
+
+    public String pizzas(String city);
+}
+----
+Here you can see that we are calling the exported method `retrievePizzeriaAddresses` on the WASM module registered with the name *pizza* in WildFly configuration and the `CityJsonSerializer`class is in charge with converting the java.lang.String city to a JSON object to be consumed by the WASM module.
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI and subsequently exposed as MCP tools, which we'll cover in the next step.
+
+
+==  Next Step: MCP Integration
+
+Well this is the simplest part as we 'just' have to use our MCP annotations and WildFly Glow will provision the MCP server for us.
+
+Note also that WildFly MCP server support is SSE only (no stdio).
+
+Our WASM service now becomes:
+[source,java]
+----
+import org.wildfly.mcp.api.Tool;
+import org.wildfly.mcp.api.ToolArg;
+import org.wildfly.wasm.api.WasmToolService;
+
+@WasmToolService(wasmToolConfigurationName = "pizza", wasmMethodName = "retrievePizzeriaAddresses", argumentSerializer = CityJsonSerializer.class)
+public interface Pizza {
+
+    @Tool(name="pizzaRetriever", description = "Get the address for the best hawaian pizzas")
+    public String pizzas(@ToolArg(description = "The city where we are looking for Hawaian pizza") String city);
+}
+
+----
+
+And now there is a new *pizzaRetriever* MCP tool available :)
+
+
+The source for this example is https://github.com/ehsavoie/wildfly-weather/tree/wasm_subsystem[here].
+
+video::oEvjWW7MBWM[youtube,width=560,height=315,title="Deploying Wasm binaries in WildFly and exposing them as MCP tools"]
```

Comment:
Is the readme https://github.com/ehsavoie/wildfly-weather/blob/ab57f0b2437654ca65df9e004757923148473d3f/README.md relevant to your example? Do I need to follow the Keycloak instructions before I can use WASM?

This PR looks good and ready to be merged as soon as you confirm that the README is correct (or is corrected :)